src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -31,109 +30,134 @@
+            this.operation = operation;
+        }
+    }
 
-            String field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
+    public String[] parseCSV(String csv){
+        // split csv with '\n' sign to get its rows
+        String[] lines = csv.split("\n");
 
-            pet.setName(field);
+        return lines;
+    }
 
-            field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
-
-            try {
-                pet.setBirthDate((new SimpleDateFormat("yyyy-MM-dd")).parse(field));
-            } catch (ParseException e) {
-                HttpHeaders headers = new HttpHeaders();
-                headers.add("errors", "date " + field + " not valid");
-                return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
+    public List<Entry> createEntries(String[] lines){
+        List<Entry> entries = new LinkedList<Entry>();
+        // now parse each field
+        for (String line: lines) {
+            // quick check if the number of ';' is correct in this line
+            if(line.count(';') != 4){
+                // print or log error
+                continue;
             }
+            // it is assumed that these functions are properly defined to check the validity of each field,
+            // and returns the parsed fields. This can also be wrapped in another function 'toEntry(line)'.
+            // some of the validity check of the original code is included, e.g., check for none or multiple pet owner
+            String[] field = line.split(";");
+            String petName = toPetName(field[0]);
+            Date date = toDate(field[1]);
+            PetType petType = toPetType(field[2]);
+            Owner owner = toOwner(field[3]);
+            String operation = toOperation(field[4]);
+            entries.add(new Entry(petName, date, petType, owner, operation));
+        }
+        return entries;
+    }
 
-            field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
-
-            if (pet != null) {
-                ArrayList<PetType> ts = (ArrayList<PetType>) clinicService.findPetTypes();
-                for (int j = 0; j < ts.size(); j++) {
-                    if (ts.get(j).getName().toLowerCase().equals(field)) {
-                        pet.setType(ts.get(j));
-                        break;
-                    }
-                }
-            }
+    public ResponseEntity<List<Pet>> importPets(@RequestBody String csv) {
 
-            field = "";
-            while (i < csv.length() && (csv.charAt(i) != ';' && csv.charAt(i) != '\n')) {
-                field += csv.charAt(i++);
-            }
+        int i = 0;
+        List<Pet> pets = new LinkedList<Pet>();
+        Pet pet;
```

Comment:
you can create that object in the for-loop - I see not need to create this outside the loop-scope?